docs/odcs.rst

```diff
@@ -24,84 +24,188 @@ idea is to not require the user to provide a yum repo url when building
 container images.  Instead, OSBS will use ODCS to generate a yum repo with the
 relevant signed packages based on user configuration.
 
+Concepts
+--------
+
+Signing Intent
+""""""""""""""
+
+When an ODCS compose is used, the idea of a signing intent is used. This refers
+to the signature of the content to be included in the container image. For
+instance, an environment may provide the following intents: ``release``,
+``beta``, and ``unsigned``. Each one of those intents is then mapped to a list
+of signing keys. These signing keys are then used during compose creation. The
+packages to be included must have been signed by either one of the keys listed.
+In the example above, the intents could be mapped to the following keys::
+
+    # Only include packages that have been signed by "my-release-key"
+    release -> my-release-key
+    # Include packages that have been signed by either "my-beta-key" or
+    # "my-release-key"
+    beta -> my-beta-key, my-release-key
+    # Do not check signature of packages - may include unsigned packages
+    unsigned -> <empty>
+
+*This mapping will be defined in reactor-config-secret which will allow updates
+to be reflected instantly for automatically triggered builds from an existing
+BuildConfig.*
+
+The definition of intents also describes the restrictive order, which will be
+enforced when building layered images. For instance, consider the case of two
+images, X and Y. Y uses X as its parent image (FROM X). If image X was built
+with "beta" intent, image Y's intent can only be "beta" or "unsigned". If the
+dist-git repo for image Y has it configured to use "release" intent, this value
+will be downgraded to "beta" at build time.
```

Comment:
I'm a bit confused by this. It seems to merge together the intent and the result, in that the developer's intent can be altered by the build system.

Why wouldn't we want this build to fail? Compare with the distribution-scope label enforcement: if the parent image's distribution scope is narrower than the image we're attempting to build, we fail.

I'm sure there is a reason for doing it the way it's documented here, but can we document what that reason is?

Also, how is this build comparing with the parent build's intent? By examining the Koji metadata for that parent build?

Reply:
> how is this build comparing with the parent build's intent? By examining the Koji metadata for that parent build?

That's correct. It'll use the Koji metadata for the parent build.

Added this paragraph to further clarify the rationale behind this behavior:
> Automatically downgrading the signing intent, instead of failing the build, is important for allowing a hierarchy of layered images to be built automatically by ImageChangeTriggers. For instance, with Continuous Integration in mind, a user may want to perform daily builds without necessarily requiring signed packages while periodically also producing builds with signed content. In this case, the signing_intent in container.yaml can be set to release for all the images in hierarchy. Whether or not the layered images in the hierarchy use signed packages can be controlled by simply overriding the signing intent of the top most image. The signing intent of the layered images would then be automatically adjusted as needed.